apps/emqx_bridge_kinesis/src/emqx_bridge_kinesis_impl_producer.erl

```diff
@@ -0,0 +1,229 @@
+    Options = proplists:get_value(config, Opts),
+    {ok, _Pid} = Result = emqx_bridge_kinesis_connector_client:start_link(Options),
+    Result.
+
+%%-------------------------------------------------------------------------------------------------
+%% Helper fns
+%%-------------------------------------------------------------------------------------------------
+
+-spec do_send_requests_sync(
+    resource_id(),
+    [{send_message, map()}],
+    state()
+) ->
+    {ok, status_code(), headers()}
+    | {ok, status_code(), headers(), body()}
+    | {error, {recoverable_error, term()}}
+    | {error, term()}.
+do_send_requests_sync(
+    InstanceId,
+    Requests,
+    #{pool_name := PoolName, templates := Templates}
+) ->
+    Result = ecpool:pick_and_do(
+        PoolName,
+        {emqx_bridge_kinesis_connector_client, query, [Requests, Templates]},
+        no_handover
+    ),
+    handle_result(Result, Requests, InstanceId).
+
+handle_result({ok, _} = Result, _Requests, _InstanceId) ->
+    Result;
+handle_result({error, {<<"ResourceNotFoundException">>, _} = Reason}, Requests, InstanceId) ->
+    ?SLOG(error, #{
+        msg => "kinesis_error_response",
+        request => Requests,
+        connector => InstanceId,
+        reason => Reason
+    }),
+    {error, {unrecoverable_error, unhealthy_target}};
+handle_result(
+    {error, {<<"ProvisionedThroughputExceededException">>, _} = Reason}, Requests, InstanceId
+) ->
+    ?SLOG(error, #{
+        msg => "kinesis_error_response",
+        request => Requests,
+        connector => InstanceId,
+        reason => Reason
+    }),
+    {error, {recoverable_error, Reason}};
+handle_result({error, {<<"InvalidArgumentException">>, _} = Reason}, Requests, InstanceId) ->
+    ?SLOG(error, #{
+        msg => "kinesis_error_response",
+        request => Requests,
+        connector => InstanceId,
+        reason => Reason
+    }),
+    {error, {unrecoverable_error, Reason}};
+handle_result({error, {econnrefused = Reason, _}}, Requests, InstanceId) ->
+    ?SLOG(error, #{
+        msg => "kinesis_error_response",
+        request => Requests,
+        connector => InstanceId,
+        reason => Reason
+    }),
+    {error, {recoverable_error, Reason}};
+handle_result({error, Reason} = Error, Requests, InstanceId) ->
+    ?SLOG(error, #{
+        msg => "kinesis_error_response",
+        request => Requests,
+        connector => InstanceId,
+        reason => Reason
+    }),
+    Error.
+
+parse_template(Config) ->
+    PayloadTemplate = #{send_message => maps:get(payload_template, Config)},
+    PartitionKeyTemplate = #{partition_key => maps:get(partition_key, Config)},
+    parse_template(maps:to_list(PayloadTemplate) ++ maps:to_list(PartitionKeyTemplate), #{}).
+
+parse_template([{Key, H} | T], Templates) ->
+    ParamsTks = emqx_placeholder:preproc_tmpl(H),
+    parse_template(
+        T,
+        Templates#{Key => ParamsTks}
+    );
+parse_template([], Templates) ->
+    Templates.
```

Comment:
```suggestion
parse_template(Config) ->
    #{ payload_template := PayloadTemplate
     , partition_key := PartitionKeyTemplate
     } = Config,
    Templates = #{send_message => PayloadTemplate, paritition_key => PartitionKeyTemplate},
    maps:map(fun(_K, V) -> emqx_placeholder:preproc_tmpl(V) end, Templates).
```

Might even save the need for the 2nd line if we use the same key name instead of `send_message`, then `maps:with/2` before `maps:map` should be enough.

Reply:
> I did't get the 2nd line saving
if we don't rename `payload_template` to `send_message`, that is.

Reply:
I got it. But I think `payload_template` makes more sense for the field name (so it would end up with a less descriptive name just to avoid 1 line). I think we could keep this extra line =)

Reply:
I also think `payload_template` is more descriptive.  That's why I'd suggest not renaming it to `send_message`.  :wink: